### PR TITLE
Add short descriptions of PeerConnection events

### LINKED ISSUE #881: 11) [rfced]  Sections 4 and 7:  For ease of the reader, please
11) [rfced]  Sections 4 and 7:  For ease of the reader, please
let us know if we may cite [W3C.webrtc] as a reminder regarding
"addTrack," "removeTrack," etc. (Section 4) as well as
"onicecandidate," "addIceCandidate," and "ontrack" (Section 7),
as follows.

Original:
 The actual API exposed in the W3C API
 may have somewhat different syntax, but should map easily to these
 concepts.
...
 More examples of SDP for WebRTC call flows, including examples with
 IPv6 addresses, can be found in [I-D.ietf-rtcweb-sdp].

Suggested:
 The actual API exposed in the W3C API [W3C.webrtc]
 may have somewhat different syntax but should map easily to these
 concepts.
...
 More examples of SDP for WebRTC call flows, including examples with
 IPv6 addresses, can be found in [SDP4WebRTC].  See [W3C.webrtc] for
 information regarding "onicecandidate," "addIceCandidate," and
 "ontrack".

### PR DESCRIPTION
**Pull request by juberti**
Fixes #881

### CODE REVIEW

**Review by ajeanmahoney — COMMENTED** [review]
It would be helpful to provide a link to the new onicecandidate section when it is first mentioned in the text.

**Review by ajeanmahoney — APPROVED** [review]

**Comment by fluffy** [comment]
Do you need to say anything about privacy implications of this or how the local addresses are filtered ?

**Comment by juberti** [comment]
@ajeanmahoney I considered providing a reference but noticed that we don't currently do this for other methods, e.g., setRemoteDescription, mentioned in the introduction. Open to ideas here but would want to be consistent.
@fluffy I could add some text similar to '... subject to the restrictions noted in <xref target="ice-candidate-policy">'

**Comment by ajeanmahoney** [comment]
I consider the request for the xref an enhancement request, not a bug. I'm okay with leaving it as-is.

**Comment by juberti** [comment]
Added text about the ICE candidate filtering and some minor editorial tweaks. @fluffy LMK if this LGTY, @mskucherawy for final approval.

File: draft-ietf-rtcweb-jsep.xml [hunk 1]
```diff
@@ -1568,6 +1584,28 @@ Suggested:
...
+          <name>onicecandidate Event</name>
+          <t>The onicecandidate event is dispatched to the application when the
+          ICE agent has discovered a new allowed local ICE candidate during ICE
+          gathering, as outlined in
+          <xref target="sec.ice-gather-overview" format="default"></xref> and 
+          subject to the restrictions discussed in 
+          <xref target="sec.ice-candidate-policy" format="default"></xref>.
+          The new candidate is supplied as an IceCandidate object 
+          (<xref target="sec.ice-candidate-format" format="default"/>) in 
+          the event, and all of its fields <bcp14>MUST</bcp14> be filled in.
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1597]: What does "filled in" mean in this context. In a number of cases these will be "null". Do you just mean they won't be "undefined"?

File: draft-ietf-rtcweb-jsep.xml [hunk 2]
```diff
@@ -1568,6 +1584,28 @@ Suggested:
...
+          <xref target="sec.ice-candidate-policy" format="default"></xref>.
+          The new candidate is supplied as an IceCandidate object 
+          (<xref target="sec.ice-candidate-format" format="default"/>) in 
+          the event, and all of its fields <bcp14>MUST</bcp14> be filled in. 
+          This candidate will also be added to the current and/or pending local
+          description according to the rules defined for Trickle ICE.</t>
+          <t>The onicecandidate event will also be dispatched to indicate
+          end-of-candidates conditions, as defined in 
+          <xref target="RFC8838" format="default"/>. For these indications,
+          the IceCandidate's candidate field <bcp14>MUST</bcp14> be omitted.
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1603]:
```suggestion
          the IceCandidate's candidate field <bcp14>MUST</bcp14> be an empty string.
```

"This carries the candidate-attribute as defined in section 15.1 of [ICE]. If this RTCIceCandidate represents an end-of-candidates indication or a peer reflexive remote candidate, candidate is an empty string. "

File: draft-ietf-rtcweb-jsep.xml [hunk 3]
```diff
@@ -1568,6 +1584,28 @@ Suggested:
...
+          The "m=" section index and MID fields may be set to signify whether
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1604]: Do you want to say "sdpMid and sdMLineIndex"?

**Review by fluffy — APPROVED** [review]
LGTM

File: draft-ietf-rtcweb-jsep.xml (hunk 1 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1597]: Well, when there is a new ICE candidate all fields will be non-null, which is what I am trying to say.

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1603]: The sec.ice-candidate-format xref should cover this, although perhaps it needs to be clarified that that xref applies to this paragraph (about eoc) as well.

File: draft-ietf-rtcweb-jsep.xml (hunk 3 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1604]: in JSEP we refer to these fields indirectly, see https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-24#section-3.5.2.1

Comment (ekr) [reply draft-ietf-rtcweb-jsep.xml:1604]: OK.

File: draft-ietf-rtcweb-jsep.xml (hunk 1 above)
Comment (ekr) [reply draft-ietf-rtcweb-jsep.xml:1597]: I might be missing something, but as I read the text raddr/rport is null for host candidates, which should trigger onicecandidate, no?

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (ekr) [reply draft-ietf-rtcweb-jsep.xml:1603]: I'm just being pedantic about "omitted" here. For JS objects that are stringy, there are three "nothing here" values:

```undefined``` [doesn't exist]
```null``` explicitly set to ```null``
```""``` explicitly set to empty string

So, what does "omitted" mean? ```undefined``` is (I think) forbidden by ```DOMString?```, but
I would usually think it meant ```null```.

File: draft-ietf-rtcweb-jsep.xml (hunk 1 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1597]: yes, but IceCandidate doesn't break those out into their own fields. IceCandidate (in JSEP) defines
- candidate (a la a=candidate)
- ufrag
- m-line index
- mid

all of which are filled in for all local candidates.

Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1597]: (as outlined in https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-24#section-3.5.2.1)

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1603]: yeah, ```null``` is what we want here. Need to see how I can make this work with the existing text.

File: draft-ietf-rtcweb-jsep.xml (hunk 1 above)
Comment (ekr) [reply draft-ietf-rtcweb-jsep.xml:1597]: Got it. I was thinking of rtcIceCandidate. Thankss for correcting me.

**Comment by juberti** [comment]
I took another pass over this to address @ekr's feedback, and also resolve an inconsistency in the document where we had previously said that all IceCandidate fields MUST be filled in.

**Comment by juberti** [comment]
PTAL

File: draft-ietf-rtcweb-jsep.xml [hunk 4]
```diff
@@ -1557,22 +1555,22 @@ Suggested:
...
-          <xref target="RFC8838" format="default"/>.</t>
+          <xref target="RFC8838" sectionFormat="comma" section="14"/>.</t>
           <t>In either case, the "m=" section index, MID, and ufrag
           fields from the supplied IceCandidate are used to determine
           which "m=" section and ICE candidate generation the
           IceCandidate belongs to, as described in
           <xref target="sec.ice-candidate-format" format="default"/> above. In the case
-          of an end-of-candidates indication, the absence of both the
+          of an end-of-candidates indication, null values for the
           "m=" section index and MID fields is interpreted to mean that
```
Comment (ajeanmahoney) [inline draft-ietf-rtcweb-jsep.xml:1571]: subject/verb agreement issue here. Perhaps "null values ... are interpreted"

**Review by ajeanmahoney — COMMENTED** [review]
I only found one typo, which I've highlighted.

File: draft-ietf-rtcweb-jsep.xml [hunk 5]
```diff
@@ -1568,6 +1582,37 @@ Suggested:
...
           candidate.</t>
         </section>
+        <section anchor="sec.onicecandidate" numbered="true" toc="default">
+          <name>onicecandidate Event</name>
+          <t>The onicecandidate event is dispatched to the application in two 
+          situations: 1) when the ICE agent has discovered a new allowed local
+          ICE candidate during ICE gathering, as outlined in
+          <xref target="sec.ice-gather-overview" format="default"></xref> and 
+          subject to the restrictions discussed in 
+          <xref target="sec.ice-candidate-policy" format="default"></xref>, and
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1592]: do you think ", or"?

File: draft-ietf-rtcweb-jsep.xml [hunk 6]
```diff
@@ -1568,6 +1582,37 @@ Suggested:
...
+          2) when an ICE gathering phase completes. The event contains a single
+          IceCandidate object, as defined in 
+          <xref target="sec.ice-candidate-format" format="default"/>.
+          </t>
+          <t>For the first case, the newly discovered candidate is reflected
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1597]: I might say "In the first case" but that's style.

File: draft-ietf-rtcweb-jsep.xml [hunk 7]
```diff
@@ -1568,6 +1582,37 @@ Suggested:
...
+          in the IceCandidate object, and all of its fields <bcp14>MUST</bcp14> be non-null.
+          This candidate will also be added to the current and/or pending local
+          description according to the rules defined for Trickle ICE.</t>
+          </t>
+          <t>For the second case, the event's IceCandidate object 
+          <bcp14>MUST</bcp14> have its candidate field set to null to indicate
+          the current gathering phase is complete, i.e., there will be no
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1604]:
```suggestion
         that the current gathering phase is complete, i.e., there will be no
```

File: draft-ietf-rtcweb-jsep.xml [hunk 8]
```diff
@@ -1568,6 +1582,37 @@ Suggested:
...
+          further onicecandidate events in this phase. However, the
+          IceCandidate's ufrag field <bcp14>MUST</bcp14> be specified to
+          indicate which ICE candidate generation is ending. The IceCandidate's 
+          "m=" section index and MID fields may be specified to indicate that
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1608]:
```suggestion
          "m=" section index and MID fields <bcp14>MAY</bcp14> be specified to indicate that
```

File: draft-ietf-rtcweb-jsep.xml [hunk 9]
```diff
@@ -1568,6 +1582,37 @@ Suggested:
...
+          the event applies to a specific "m=" section, or set to null to 
+          indicate it applies to all "m=" sections in the current ICE candidate
+          generation. This event can be used to generate an end-of-candidates
+          indication, as defined in 
+          <xref target="RFC8838" sectionFormat="comma" section="13"/>.
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml:1613]:
```suggestion
          generation. This event can be used by the JavaScript to generate an end-of-candidates
          indication, as defined in 
          <xref target="RFC8838" sectionFormat="comma" section="13"/>. 
```

**Review by ekr — APPROVED** [review]
Generally LGTM

File: draft-ietf-rtcweb-jsep.xml (hunk 5 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1592]: agreed

File: draft-ietf-rtcweb-jsep.xml (hunk 6 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1597]: sg

File: draft-ietf-rtcweb-jsep.xml (hunk 9 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1613]: good call, tweaked the wording a bit

File: draft-ietf-rtcweb-jsep.xml (hunk 4 above)
Comment (juberti) [reply draft-ietf-rtcweb-jsep.xml:1571]: good catch

**Comment by juberti** [comment]
all comments addressed. Will plan to merge once Travis passes unless I hear otherwise.

**Review by fluffy — APPROVED** [review]
LGTM

**Pull request merged**